### PR TITLE
Implementation of the Material Date Range Picker.

### LINKED ISSUE #50618: Natively support a date range picker
## Use case

Date pickers are great, but sometimes users need to pick a _range_ of dates spanning days, weeks, months, or even years. 

It avoids having to use two fields like so:

Screenshot from 2020-02-11 20-55-58
(_Yes I know it's the same date! It's just to show the two fields._)

Why would you use _two_ fields when you could use just _one_?

## Proposal

There's packages for date range pickers (https://pub.dev/packages/date_range_picker, https://pub.dev/packages/flutter_date_pickers) but it's not *official* and it's a part of the [material design specifications](https://material.io/components/pickers/#anatomy). Android has this functionality so should Flutter. This should be natively supported in Flutter and as far as I can tell, it isn't.

(Source: [Pickers](https://material.io/components/pickers/#anatomy))

### PR DESCRIPTION
**Pull request by darrenaustin**
## Description

This PR adds support for the [Material Date Range Picker](https://material.io/components/pickers/#mobile-pickers) to Flutter's Material library.

DateRangePicker

This includes support for both the full screen modal dialog with a scrolling calendar grid, as well as a text input mode that has two input fields. The new API consists mostly of the function:

```dart
Future<DateTimeRange> showDateRangePicker({
  @required BuildContext context,
  DateTimeRange initialDateRange,
  @required DateTime firstDate,
  @required DateTime lastDate,
  DateTime currentDate,
  DatePickerEntryMode initialEntryMode = DatePickerEntryMode.calendar,
  String helpText,
  String cancelText,
  String confirmText,
  String saveText,
  Locale locale,
  bool useRootNavigator = true,
  RouteSettings routeSettings,
  TextDirection textDirection,
  TransitionBuilder builder,
  String errorFormatText,
  String errorInvalidText,
  String errorInvalidRangeText,
  String fieldStartHintText,
  String fieldEndHintText,
  String fieldStartLabelText,
  String fieldEndLabelText,
}) async
```

Which is similar to `showDatePicker`, but returns a new `DateTimeRange` that describes the start and end dates of the range.

Kudos to @rami-a, upon whose work this is based. 

## Related Issues

Fixes: #50618

## Tests

Added a new `date_range_picker_test.dart` file with tests for the various features of the new picker.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them?

- [x] No, no existing tests failed, so this is *not* a breaking change.

### CODE REVIEW

File: packages/flutter/lib/src/material/pickers/date_range_picker_dialog.dart
Comment (rami-a) [inline packages/flutter/lib/src/material/pickers/date_range_picker_dialog.dart:109]: nit: Should all the string params be grouped together?

File: packages/flutter/lib/src/material/pickers/date_range_picker_dialog.dart
Comment (gspencergoog) [inline packages/flutter/lib/src/material/pickers/date_range_picker_dialog.dart:38]: Should it be "Material Design"?  It's a proper noun, so maybe.

File: packages/flutter/lib/src/material/pickers/date_range_picker_dialog.dart
Comment (gspencergoog) [inline packages/flutter/lib/src/material/pickers/date_range_picker_dialog.dart:41]:
```suggestion
/// the user saves their selection. If the user cancels the modal dialog, null is
```

File: packages/flutter/lib/src/material/pickers/date_range_picker_dialog.dart
Comment (gspencergoog) [inline packages/flutter/lib/src/material/pickers/date_range_picker_dialog.dart:49]: I think we always use "must" when we talk about nullability: e.g. "Both must be non-null" or "Both must not be null".

File: packages/flutter/lib/src/material/pickers/date_range_picker_dialog.dart
Comment (gspencergoog) [inline packages/flutter/lib/src/material/pickers/date_range_picker_dialog.dart:68]: I'd add definite articles here and below:
```suggestion
///   * [helpText], the label displayed at the top of the dialog.
```

File: packages/flutter/lib/src/material/pickers/date_range_picker_dialog.dart
Comment (gspencergoog) [inline packages/flutter/lib/src/material/pickers/date_range_picker_dialog.dart:74]: Long lines of doc comments should be wrapped at 80 columns.

File: packages/flutter/lib/src/material/pickers/date_range_picker_dialog.dart
Comment (rami-a) [inline packages/flutter/lib/src/material/pickers/date_range_picker_dialog.dart:351]: You can use `BorderRadius.zero` here

File: packages/flutter/lib/src/material/pickers/date_range_picker_dialog.dart
Comment (rami-a) [inline packages/flutter/lib/src/material/pickers/date_range_picker_dialog.dart:401]: There will be theming support coming in the future right? We'd want to be able to customize the corner radius for example.

File: packages/flutter/lib/src/material/pickers/date_range_picker_dialog.dart
Comment (rami-a) [inline packages/flutter/lib/src/material/pickers/date_range_picker_dialog.dart:493]: Should probably be consistent with usage of `.0` for doubles

File: packages/flutter/lib/src/material/pickers/date_range_picker_dialog.dart
Comment (gspencergoog) [inline packages/flutter/lib/src/material/pickers/date_range_picker_dialog.dart:473]: You could probably format lines like this like so to be closer to 80 chars:
```suggestion
    final TextStyle startDateStyle = headlineStyle?.apply(
      color: selectedStartDate != null ? headerForeground : headerDisabledForeground,
    );
```

File: packages/flutter/lib/src/material/pickers/date_range_picker_dialog.dart
Comment (gspencergoog) [inline packages/flutter/lib/src/material/pickers/date_range_picker_dialog.dart:606]: This inline conditional chain is getting a little out of hand... Maybe use a conditional block or two here?

File: packages/flutter/test/material/date_range_picker_test.dart
Comment (rami-a) [inline packages/flutter/test/material/date_range_picker_test.dart:95]: Should use a different string for save text than what the default is

File: packages/flutter/lib/src/material/pickers/date_utils.dart
Comment (gspencergoog) [inline packages/flutter/lib/src/material/pickers/date_utils.dart:131]: When you say "appropriate", you mean "locale-appropriate", right? Not just, like, without cuss words. :-)

File: packages/flutter/lib/src/material/pickers/input_date_picker.dart
Comment (gspencergoog) [inline packages/flutter/lib/src/material/pickers/input_date_picker.dart:251]: Why not private?

File: packages/flutter/lib/src/material/pickers/input_date_picker.dart
Comment (gspencergoog) [inline packages/flutter/lib/src/material/pickers/input_date_picker.dart:264]:
```suggestion
  /// separator using one of these characters (slash, period, comma, dash, or space).
```

File: packages/flutter/lib/src/material/pickers/input_date_picker.dart
Comment (gspencergoog) [inline packages/flutter/lib/src/material/pickers/input_date_picker.dart:270]: super-meta nit: doesn't _have_ to be a doc comment.  Fine to leave it though.

File: packages/flutter/lib/src/material/pickers/input_date_range_picker.dart
Comment (gspencergoog) [inline packages/flutter/lib/src/material/pickers/input_date_range_picker.dart:17]:
```suggestion
/// end dates that represent a range of dates.
```

**Review by rami-a — COMMENTED** [review]
Code looks great, just a few nits and comments. Additionally found a few UI issues outlined below:

- The full screen modal should extend to the safe area
Simulator Screen Shot - iPhone 11 Pro - 2020-04-29 at 10 14 40
Simulator Screen Shot - iPhone 11 Pro - 2020-04-29 at 10 22 15


- The selected range text should go to 2 lines on the input mode in landscape mode.
Simulator Screen Shot - iPhone 11 Pro - 2020-04-29 at 10 18 22

- Picker overflows when keyboard is up in landscape
Simulator Screen Shot - iPhone 11 Pro - 2020-04-29 at 10 20 26

File: packages/flutter/lib/src/material/pickers/input_date_range_picker.dart
Comment (gspencergoog) [inline packages/flutter/lib/src/material/pickers/input_date_range_picker.dart:166]:
```suggestion
  /// Will return true if the range is valid. If not, it will
```

File: packages/flutter/lib/src/material/pickers/date_range_picker_dialog.dart
Comment (rami-a) [inline packages/flutter/lib/src/material/pickers/date_range_picker_dialog.dart:424]: Dialog defaults to 24, I wonder if this should be left out so that when a user modifies the `DialogTheme`'s elevation, it applies to the pickers too (same comment probably applies to the already merged DatePicker)

Comment (darrenaustin) [reply packages/flutter/lib/src/material/pickers/date_range_picker_dialog.dart:424]: Yeah, I just caught this when I was looking at the shape. I will fix this shortly and double check the Date Picker (but I am pretty sure I already deferred this to theme for that one). Thx.

File: packages/flutter/lib/src/material/pickers/calendar_date_range_picker.dart
Comment (darrenaustin) [reply packages/flutter/lib/src/material/pickers/calendar_date_range_picker.dart:23]: It was mostly because I wasn't yet ready to commit to it being part of the public API, but didn't want to add it's bulk to the already long `date_range_picker_dialog.dart`. I really wish dart had a 'package private' access control for things like this.

File: packages/flutter/lib/src/material/pickers/input_date_picker.dart
Comment (darrenaustin) [reply packages/flutter/lib/src/material/pickers/input_date_picker.dart:251]: Because it is shared between the `InputDatePickerFormField` and the `InputDateRangePicker` which is in another file. I could move the range picker into the `input_date_picker.dart` file, but `InputDateRangePicker` is not a public API either.

Again a package private feature would be welcome. I didn't see many examples of how to handle this sort of situation elsewhere in the framework. Most components are as complicated as these date pickers, so they just put everything in one large files. With the date pickers that just seemed too unwieldy.

File: packages/flutter/lib/src/material/pickers/calendar_date_range_picker.dart
Comment (gspencergoog) [reply packages/flutter/lib/src/material/pickers/calendar_date_range_picker.dart:23]: Well, it does have "library private", but the library can only be extended to multiple files with the `part of` directive, which we discourage. I know what you mean though: it would be nice if things could be made private to anything in the pub package.

Like it or not, it's part of the public API, it just isn't exported with the top level include points.  Anyone can use this class with:
```dart
import 'package:/flutter/src/material/pickers/calendar_date_range_picker.dart';
```

Sure, it's not likely, but they could.

Anyhow, it's not a big deal: I suspect that a) people are on their own if they do that anyhow, and 2) that we might want to make this a public API anyhow.

If you're not ready to make it public, though, it's probably OK to put it in the other huge file and make it private. We have plenty of those in Flutter (mostly for that reason) already. `widgets/basic.dart` is 7144 lines...

Comment (darrenaustin) [reply packages/flutter/lib/src/material/pickers/calendar_date_range_picker.dart:23]: Yeah, I mostly landed on "a) people are on their own if they do that". It is called out as a "DON'T" in the effective dart guide (https://dart.dev/guides/language/effective-dart/usage#dont-import-libraries-that-are-inside-the-src-directory-of-another-package), but I totally get the point that it is technically still public.

I am leaning toward leaving this way for now. That will make it slightly easier if we do want to make it public at some point.

File: packages/flutter/lib/src/material/pickers/date_range_picker_dialog.dart
Comment (darrenaustin) [reply packages/flutter/lib/src/material/pickers/date_range_picker_dialog.dart:417]: Yup, I updated this to defer to the shape from the theme if there is one.

**Comment by darrenaustin** [comment]
> Code looks great, just a few nits and comments. Additionally found a few UI issues outlined below:

Thanks @rami-a, I just pushed changes that I think address everything you brought up except for the landscape input dialog overflowing. Still looking into that one. What device where you seeing it on?

**Comment by rami-a** [comment]
> > Code looks great, just a few nits and comments. Additionally found a few UI issues outlined below:
> 
> Thanks @rami-a, I just pushed changes that I think address everything you brought up except for the landscape input dialog overflowing. Still looking into that one. What device where you seeing it on?

I was seeing it on the iPhone 11 pro simulator, though I also saw it on the regular date picker under the same conditions.

**Comment by rami-a** [comment]
> > > Code looks great, just a few nits and comments. Additionally found a few UI issues outlined below:
> > 
> > 
> > Thanks @rami-a, I just pushed changes that I think address everything you brought up except for the landscape input dialog overflowing. Still looking into that one. What device where you seeing it on?
> 
> I was seeing it on the iPhone 11 pro simulator, though I also saw it on the regular date picker under the same conditions.

One possible solution, could be to put the header into a `SingleChildScrollView` to allow for it to scroll in landscape?

**Comment by rami-a** [comment]
Found one more small issue, the Picker Dialog's content should respect the safe area.

Simulator Screen Shot - iPhone 11 Pro - 2020-05-04 at 10 35 15

**Review by rami-a — APPROVED** [review]
LGTM once those last 2 issues are fixed

**Comment by darrenaustin** [comment]
> Found one more small issue, the Picker Dialog's content should respect the safe area.

Didn't you just complain that the dialog should cover the safe area? :smirk:

Kidding aside, I am a little confused at the expectations here. Is it just the content area that should honor the safe area, but the dialog as a whole shouldn't? Or is it just the bottom safe area that is of concern?

**Comment by darrenaustin** [comment]
> One possible solution, could be to put the header into a `SingleChildScrollView` to allow for it to scroll in landscape?

Possibly. I have been playing with this for a while, and the main problem is that on iOS the keyboard that we bring up for `TextInputType.datetime` is the full alphanumeric keyboard that fills over half the vertical space on a iPhone 11 in landscape. That leads to very little space for us. We could switch over to the numeric keypad which is shorter, but it lacks any separators needed for dates. I may try to put it in a scroll view if I can't find another way to make it fit.

**Comment by rami-a** [comment]
> > Found one more small issue, the Picker Dialog's content should respect the safe area.
> 
> Didn't you just complain that the dialog should cover the safe area? 😏
> 
> Kidding aside, I am a little confused at the expectations here. Is it just the content area that should honor the safe area, but the dialog as a whole shouldn't? Or is it just the bottom safe area that is of concern?

haha yes I realize that it sounds like I'm changing my mind. But yeah the entire dialog should extend into the safe area so that you can't see the page behind it. However the content area should honor the safe area, should be as simple as wrapping the Scaffold's body with a `SafeArea`

File: packages/flutter/lib/src/material/pickers/calendar_date_range_picker.dart
Comment (HansMuller) [inline packages/flutter/lib/src/material/pickers/calendar_date_range_picker.dart:44]: All of the other parameters can be null too? 

If they're not null, assert that lastDate >= firstDate etc?

File: packages/flutter/lib/src/material/pickers/calendar_date_range_picker.dart
Comment (HansMuller) [inline packages/flutter/lib/src/material/pickers/calendar_date_range_picker.dart:82]: Should be a companion dispose() override that calls _controller.dispose()

File: packages/flutter/lib/src/material/pickers/calendar_date_range_picker.dart
Comment (HansMuller) [inline packages/flutter/lib/src/material/pickers/calendar_date_range_picker.dart:208]: NICE

File: packages/flutter/lib/src/material/pickers/calendar_date_range_picker.dart
Comment (HansMuller) [inline packages/flutter/lib/src/material/pickers/calendar_date_range_picker.dart:383]: const?

File: packages/flutter/lib/src/material/pickers/calendar_date_range_picker.dart
Comment (HansMuller) [inline packages/flutter/lib/src/material/pickers/calendar_date_range_picker.dart:554]: I've tried to only use `=>` for one-liners that return value. So 
```dart
onTap: () { onChanged(dayToBuild); }
```

File: packages/flutter/lib/src/material/pickers/calendar_date_range_picker.dart
Comment (HansMuller) [inline packages/flutter/lib/src/material/pickers/calendar_date_range_picker.dart:613]: this can go on the previous line

**Review by HansMuller — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/material/pickers/calendar_date_range_picker.dart
Comment (darrenaustin) [reply packages/flutter/lib/src/material/pickers/calendar_date_range_picker.dart:404]: Unfortunately the asserts use DateTime methods which prevent this. I could move them to the body of the constructor, but as this was an internal API, it didn't seem worth it.

Comment (HansMuller) [reply packages/flutter/lib/src/material/pickers/calendar_date_range_picker.dart:404]: The assert provide more safety than the const provides readability. Sounds OK.

File: packages/flutter/lib/src/material/pickers/input_date_picker.dart
Comment (onemanstartup) [inline packages/flutter/lib/src/material/pickers/input_date_picker.dart:229]: @darrenaustin Why is this hardcoded? If I have InputDecorationTheme with filled: false, I expect it to be false, but now it's not the case.

Comment (darrenaustin) [reply packages/flutter/lib/src/material/pickers/input_date_picker.dart:229]: It is following the Material Design for this picker, which calls for it to be filled. That said, we should probably defer to the ambient `InputDecorationTheme` if one is set for this. We have a future theming pass that we will be doing on the date pickers that will allow for more customization, but I might be able to put together a small PR to fix this one issue.

Comment (onemanstartup) [reply packages/flutter/lib/src/material/pickers/input_date_picker.dart:229]: @darrenaustin Thanks. It would be great for my use case.

**Comment by jhionan** [comment]
Why not exporting the CalendarDateRangePicker class, like CalendarDatePicker, so we can create our custom date range picker, as it was merged we don't have access to the widget, only for the showRangeDatePicker() function. I with to customize the picker as my design did. I cant see any reason why not puting the CalendarDateRangePicker class exported at public api too.

**Comment by darrenaustin** [comment]
Hi @jhionan 
> Why not exporting the CalendarDateRangePicker class, like CalendarDatePicker, so we can create our custom date range picker

I did consider this, but time constraints didn't allow me to make this component ready for exposure as a publicly supported API. It is definitely on my list of things to revisit. Please open a feature request issue so we can track this. Thx.

**Comment by jhionan** [comment]
I exported your widget for my use, and its very functional as it is, in my case I just want the CalendarRangeDatePicker class. The only suggestion I have for a public api is making it easier to change colors. I know it can be done by changing the theme, but I feel that would be great if it also have a straight forward color selection.

And good job!

**Pull request merged**
